Make SpeechSynthesis an ActiveDOMObject

szewai opened the pull request:
#### 5eb9dd27dc3df7998ca679a50fd2b4e27680378e
<pre>
Make SpeechSynthesis an ActiveDOMObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=260207">https://bugs.webkit.org/show_bug.cgi?id=260207</a>

Reviewed by Ryosuke Niwa.

Keep wrapper of SpeechSynthesis alive when voiceschanged event might be fired and it has event handler.

Test: fast/speechsynthesis/speech-synthesis-voiceschanged-gc.html.

* LayoutTests/fast/speechsynthesis/speech-synthesis-voiceschanged-gc-expected.txt: Added.
* LayoutTests/fast/speechsynthesis/speech-synthesis-voiceschanged-gc.html: Added.
* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::Ref&lt;SpeechSynthesis&gt;SpeechSynthesis::create):
(WebCore::SpeechSynthesis::SpeechSynthesis):
(WebCore::SpeechSynthesis::setPlatformSynthesizer):
(WebCore::SpeechSynthesis::voicesDidChange):
(WebCore::SpeechSynthesis::getVoices):
(WebCore::SpeechSynthesis::resumeSynthesis):
(WebCore::SpeechSynthesis::simulateVoicesListChange):
(WebCore::SpeechSynthesis::activeDOMObjectName const):
(WebCore::SpeechSynthesis::virtualHasPendingActivity const):
(WebCore::SpeechSynthesis::eventListenersDidChange):
(WebCore::SpeechSynthesis::resume): Deleted.
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/Modules/speech/SpeechSynthesis.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::simulateSpeechSynthesizerVoiceListChange):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/267385@main">https://commits.webkit.org/267385@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce2e99b547e429f8b3563c468b9547b5128023eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17130 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18151 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16566 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17737 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17004 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18920 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14248 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21646 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15235 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15006 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19321 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15591 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13247 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14805 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14681 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3942 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19173 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15418 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/speechsynthesis/speech-synthesis-voiceschanged-gc.html:
```diff
@@ -0,0 +1,32 @@
+        testRunner.waitUntilDone();
+
+    const voices = speechSynthesis.getVoices();
+    speechSynthesis.onvoiceschanged = () => { finishJSTest(); };
```
We should probably add an expando property on speechSynthesis or store speechSynthesis in a [WeakRef](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) and check that the wrapper is identical before/after GC.

szewai replied:
Ok, will add an expando property.

rniwa commented on Source/WebCore/Modules/speech/SpeechSynthesis.h:
```diff
@@ -88,32 +89,39 @@ class SpeechSynthesis : public PlatformSpeechSynthesizerClient, public SpeechSyn
     bool m_isPaused;
     BehaviorRestrictions m_restrictions;
     WeakPtr<SpeechSynthesisClient> m_speechSynthesisClient;
+    std::atomic<bool> m_hasEventListener;
```
Why does this need to be atomic?

szewai replied:
`virtualHasPendingActivity()` and `eventListenersDidChange()` are invoked on different threads (`virtualHasPendingActivity()` is on a non-main thread).

rniwa replied:
Yeah but `virtualHasPendingActivity` gets called while the main thread is paused so I don't think there is a chance of data race here.

szewai replied:
I see; will update it

rniwa commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1571,6 +1571,20 @@ ExceptionOr<void> Internals::setFormControlStateOfPreviousHistoryItem(const Vect
+        return;
+    }
+
+    auto document = contextDocument();
```
Please use RefPtr here.

szewai replied:
Sure.

rniwa commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1571,6 +1571,20 @@ ExceptionOr<void> Internals::setFormControlStateOfPreviousHistoryItem(const Vect
+    if (!document || !document->domWindow())
+        return;
+
+    if (auto synthesis = LocalDOMWindowSpeechSynthesis::speechSynthesis(*document->domWindow()))
```
and here.

szewai replied:
Sure.

rniwa commented on Source/WebCore/Modules/speech/SpeechSynthesis.h:
```diff
@@ -88,32 +89,39 @@ class SpeechSynthesis : public PlatformSpeechSynthesizerClient, public SpeechSyn
     bool m_isPaused;
     BehaviorRestrictions m_restrictions;
     WeakPtr<SpeechSynthesisClient> m_speechSynthesisClient;
+    bool m_hasEventListener;
```
Please initialize this variable to false.

szewai committed:
Make SpeechSynthesis an ActiveDOMObject
https://bugs.webkit.org/show_bug.cgi?id=260207

Reviewed by Ryosuke Niwa.

Keep wrapper of SpeechSynthesis alive when voiceschanged event might be fired and it has event handler.

Test: fast/speechsynthesis/speech-synthesis-voiceschanged-gc.html.

* LayoutTests/fast/speechsynthesis/speech-synthesis-voiceschanged-gc-expected.txt: Added.
* LayoutTests/fast/speechsynthesis/speech-synthesis-voiceschanged-gc.html: Added.
* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::Ref<SpeechSynthesis>SpeechSynthesis::create):
(WebCore::SpeechSynthesis::SpeechSynthesis):
(WebCore::SpeechSynthesis::setPlatformSynthesizer):
(WebCore::SpeechSynthesis::voicesDidChange):
(WebCore::SpeechSynthesis::getVoices):
(WebCore::SpeechSynthesis::resumeSynthesis):
(WebCore::SpeechSynthesis::simulateVoicesListChange):
(WebCore::SpeechSynthesis::activeDOMObjectName const):
(WebCore::SpeechSynthesis::virtualHasPendingActivity const):
(WebCore::SpeechSynthesis::eventListenersDidChange):
(WebCore::SpeechSynthesis::resume): Deleted.
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/Modules/speech/SpeechSynthesis.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::simulateSpeechSynthesizerVoiceListChange):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/267385@main
